### PR TITLE
Loki: backend: add error-prop if loki parse error happened

### PR DESCRIPTION
**Pull request by gabor**
imagine you have logs in loki, and most are in JSON format, but some are not.
if you run a query like `{...} | json`, then the json-transform will fail for some of the log-lines. Loki indicates it by adding a label named `error` to that log-line.

we detect this in grafana, and in the logs-visualization we show an error-message saying `Error when parsing some of the logs`.

this pull request adds this to the backend-mode.

NOTE: also, in frontend-mode this was broken currently, so this PR fixes it in frontend-mode too.

how to test:
1. make sure you have the `lokiBackendMode` feature flag enabled
2. go to grafana-explore
3. have logs in not-json format, and use the `{...} | json` query, or, the same for `logfmt`. (have logs in not-logfmt format, and use the `{...} | logfmt` query
4. you should see the message "Error when parsing some of the logs" above the logs, to the left, in red.
5. disable the `lokiBackendMode` feature flag, and do the test again, the error message should appear

### CODE REVIEW

File: public/app/plugins/datasource/loki/backendResultTransformer.test.ts [hunk 1]
```diff
@@ -31,7 +31,10 @@ const inputFrame: DataFrame = {
           json: true,
         },
       },
-      values: new ArrayVector(['{ "level": "info", "code": "41🌙" }', '{ "level": "error", "code": "41🌙" }']),
+      values: new ArrayVector([
+        { level: 'info', code: '41🌙' },
+        { level: 'error', code: '41🌙' },
+      ]),
```
Comment (gabor) [inline public/app/plugins/datasource/loki/backendResultTransformer.test.ts:37]: this unrelated test-case had a bug in it, i fixed it (switched from JSON_IN_STRING to javascript-object)

File: public/app/plugins/datasource/loki/result_transformer.ts [hunk 2]
```diff
@@ -346,8 +346,16 @@ export function lokiStreamsToDataFrames(
...
+  const labelsField = dataFrame.fields.find((f) => f.name === 'labels');
+
+  if (meta.custom && labelsField) {
+    const allLabelNames = labelsField.values
+      .toArray()
+      .map((labelObject) => Object.keys(labelObject))
+      .flat();
+    if (allLabelNames.some((l) => l === '__error__')) {
+      meta.custom.error = 'Error when parsing some of the logs';
+    }
```
Comment (gabor) [inline public/app/plugins/datasource/loki/result_transformer.ts:358]: this was broken in frontend-mode, because we switched the dataframe-format in the past, but this code did not get updated correctly. as you can see, it's the difference between taking labels from the field's `.labels` prop, and taking data from the labels-field itself.

File: public/app/plugins/datasource/loki/backendResultTransformer.test.ts [hunk 3]
```diff
@@ -39,12 +39,11 @@ const inputFrame: DataFrame = {
...
-        custom: {
-          json: true,
-        },
-      },
-      values: new ArrayVector(['{ "level": "info", "code": "41🌙" }', '{ "level": "error", "code": "41🌙" }']),
+      config: {},
+      values: new ArrayVector([
+        { level: 'info', code: '41🌙' },
+        { level: 'error', code: '41🌙' },
+      ]),
```
Comment (gabor) [inline public/app/plugins/datasource/loki/backendResultTransformer.test.ts:46]: this unrelated change just fixes an unrelated test-case (data-format changed in the past but i forgot to update the test case)

**Review by joey-grafana — APPROVED** [review]
LGTM 👍 

The only thing I find a little bit strange is that we have an error at the top 'Error when parsing some of the logs', yet we have a JSON warning on each line. Seems likes a good approach now, however may need to revisit the error showing pattern in the future.

**Comment by gabor** [comment]
@joey-grafana 
> The only thing I find a little bit strange is that we have an error at the top 'Error when parsing some of the logs', yet we have a JSON warning on each line. Seems likes a good approach now, however may need to revisit the error showing pattern in the future.

yes, i agree.  for now i just want to be "the same as before" 😄 , but it's definitively something that may need to be revisited from an UX perspective.
